migrations/versions/e88e0781857f_.py

```diff
@@ -0,0 +1,38 @@
+"""Adding UniqueConstraint to table nicety, deleting duplicate rows
+
+Revision ID: e88e0781857f
+Revises: 490ca3791aac
+Create Date: 2021-05-10 16:20:13.973239
+
+"""
+from alembic import op
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = 'e88e0781857f'
+down_revision = '490ca3791aac'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_unique_constraint(None, 'nicety', ['author_id', 'target_id', 'end_date'])
+    # ### end Alembic commands ###
+
+    connection = op.get_bind()
+    result = connection.execute('''select A.id, B.id from nicety as A
+    join nicety as B
+    on A.author_id = B.author_id
+    and A.target_id = B.target_id
+    and A.text = B.text
+    and A.id > B.id;''')
+    for row in result:
+        print(row[0])
+        # One option for deletion
+        op.execute('delete from nicety where id = {0};'.format(row[0]))
```

Comment:
A rule I've adopted in working with SQL databases is to never, ever, do string formatting with parameters. That practice is the root of [SQL injection](https://en.wikipedia.org/wiki/SQL_injection) vulnerabilities, which is such a significant security problem that I think it's worth altering habits to avoid even feeling the need to consider the question "right, but is that safe to do in THIS instance?" (it probably is here, but let's still please avoid it!)

The way to prevent SQL injection attacks is to use [bound parameters](https://en.wikipedia.org/wiki/Prepared_statement). Alembic + SQLAlchemy make that slightly difficult here, but basically they want us to use their expression API rather than their SQL API:

```suggestion
        op.execute(sa.delete(nicety).where(nicety.c.id == row[0]))
```

In order to do that, we need to [re-define](https://stackoverflow.com/a/45805092) the relevant bit of schema, as it exists before the migration:

```py
from sqlalchemy.sql import table, column
nicety = table('nicety',
    column('id', sa.Integer),
)
```

Reply:
Thank you for catching all these issues! Will fix and resubmit.